src/serialize.h

```diff
@@ -375,6 +378,160 @@ I ReadVarInt(Stream& is)
 #define VARINT(obj) REF(WrapVarInt(REF(obj)))
 #define LIMITED_STRING(obj,n) REF(LimitedString< n >(REF(obj)))
 
+class CMFToken
+{
+public:
+    CMFToken() : tag(0), format(PositiveNumber) {}
+    CMFToken(uint32_t tag, const uint256& inData)
+        : tag(tag),
+        format(ByteArray)
+    {
+        std::vector<char> tmp;
+        tmp.resize(32);
+        std::copy(inData.begin(), inData.end(), tmp.begin());
+        data = tmp;
+    }
+    explicit CMFToken(uint32_t tag, bool in = true) : tag(tag), format(in ? BoolTrue : BoolFalse) {}
+    CMFToken(uint32_t tag, int32_t in) : tag(tag), format(in >= 0 ? PositiveNumber : NegativeNumber), data(in) {}
+    CMFToken(uint32_t tag, uint64_t in) : tag(tag), format(PositiveNumber), data(in) {}
+    CMFToken(uint32_t tag, std::string in) : tag(tag), format(String), data(in) {}
+    CMFToken(uint32_t tag, std::vector<char> in) : tag(tag), format(ByteArray), data(in) {}
+    typedef boost::variant<int32_t, bool, uint64_t, std::string, std::vector<char>, double> variant;
+    enum Format {
+        PositiveNumber = 0,
+        NegativeNumber,
+        String,
+        ByteArray,
+        BoolTrue,
+        BoolFalse,
+        Double
+    };
+    uint32_t tag;
+    Format format;
+    variant data;
+
+    int intData() {
+        if (data.which() == 0)
+            return boost::get<int32_t>(data);
+        else if (data.which() == 2)
+            return boost::get<uint64_t>(data);
+        return 0;
+    }
+    uint64_t longData() {
+        if (data.which() == 0)
+            return boost::get<int32_t>(data);
+        else if (data.which() == 2)
+            return boost::get<uint64_t>(data);
+        return 0;
+    }
+
+    std::vector<unsigned char> unsignedByteArray() { // provided to be compatible with silly code that uses unsigned chars.
+        auto plainData = boost::get<std::vector<char> >(data);
+        std::vector<unsigned char> bytes; // using unsigned char is really bad practice :(
+        bytes.resize(plainData.size());
+        memcpy(&bytes[0], &plainData[0], plainData.size());
+        return bytes;
+    }
+
+    template<typename Stream>
+    void Unserialize(Stream& s, int type, int version)
+    {
+        tag = ser_readdata8(s);
+        format = static_cast<Format>(tag & 7);
+        tag = tag >> 3;
+        if (tag == 31)
+            tag = ReadVarInt<Stream, uint32_t>(s);
+
+        switch (format) {
+        case NegativeNumber:
+            data = ReadVarInt<Stream, int32_t>(s) * -1;
+            break;
+        case PositiveNumber: {
+            int64_t tmp = ReadVarInt<Stream, int64_t>(s);
+            if (tmp > 0xFFFFFF)
+                data = (uint64_t) tmp;
+            else
+                data = (int32_t) tmp;
```

Comment:
What's the reason for preferring C-style cast over C++? Is it just for brevity? Also why the magic number `0xFF FF FF`? Why not `0x7F FF FF FF` or use `numeric_limits<int32_t>::max()`?